From d911278f47089e710a0a6335c6a81c4c222e3760 Mon Sep 17 00:00:00 2001
From: Amit Kumar Mahapatra <amit.kumar-mahapatra@xilinx.com>
Date: Sun, 26 Jan 2020 13:46:23 -0700
Subject: [PATCH 1033/1852] mtd: spi-nor: Added axi-qspi support in spi-nor
 framework

commit 24bcc7359bdd609cedf0a715d77dc39e4ad782db from
https://github.com/Xilinx/linux-xlnx.git

For axi-qspi update spi-nor framework to set address width to 3byte,
disable 4byte mode and use ear to access flash memory greater than 16MB.

Signed-off-by: Amit Kumar Mahapatra <amit.kumar-mahapatra@xilinx.com>
Signed-off-by: Naga Sureshkumar Relli <nagasure@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/mtd/spi-nor/spi-nor.c | 20 ++++++++++++++------
 1 file changed, 14 insertions(+), 6 deletions(-)

diff --git a/drivers/mtd/spi-nor/spi-nor.c b/drivers/mtd/spi-nor/spi-nor.c
index 782b4e0bd868..37d90d8c987a 100644
--- a/drivers/mtd/spi-nor/spi-nor.c
+++ b/drivers/mtd/spi-nor/spi-nor.c
@@ -5193,13 +5193,21 @@ static int spi_nor_set_addr_width(struct spi_nor *nor)
 			    nor->info->flags & SPI_NOR_4B_OPCODES) {
 				spi_nor_set_4byte_opcodes(nor);
 			} else {
-				nor->params.set_4byte(nor, true);
-				if (nor->isstacked) {
-					nor->spi->master->flags |=
-						SPI_MASTER_U_PAGE;
+				np_spi = of_get_next_parent(np);
+				if (of_property_match_string(np_spi,
+							     "compatible",
+							     "xlnx,xps-spi-2.00.a") >= 0) {
+					nor->addr_width = 3;
+					nor->params.set_4byte(nor, false);
+				} else {
 					nor->params.set_4byte(nor, true);
-					nor->spi->master->flags &=
-						~SPI_MASTER_U_PAGE;
+					if (nor->isstacked) {
+						nor->spi->master->flags |=
+							SPI_MASTER_U_PAGE;
+						nor->params.set_4byte(nor, true);
+						nor->spi->master->flags &=
+							~SPI_MASTER_U_PAGE;
+					}
 				}
 			}
 #ifdef CONFIG_OF
-- 
2.31.1

